src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
-                     HANDLE* const hInput,              // _Out_
-                     HANDLE* const hOutput,             // _Out_
-                     HANDLE* const hSignal,             // _Out_
-                     PROCESS_INFORMATION* const piPty); // _Out_
+// A case-insensitive wide-character map is used to store environment variables
+// due to documented requirements:
+//
+//      "All strings in the environment block must be sorted alphabetically by name.
+//      The sort is case-insensitive, Unicode order, without regard to locale.
+//      Because the equal sign is a separator, it must not be used in the name of
+//      an environment variable."
+//      https://docs.microsoft.com/en-us/windows/desktop/ProcThread/changing-environment-variables
+//
+struct WStringCaseInsensitiveCompare
+{
+    bool operator()(const std::wstring& lhs, const std::wstring& rhs) const
+    {
+        return (::_wcsicmp(lhs.c_str(), rhs.c_str()) < 0);
+    }
+};
+
+using EnvironmentVariableMapW = std::map<std::wstring, std::wstring, WStringCaseInsensitiveCompare>;
+
+HRESULT CreateConPty(const std::wstring& cmdline,                       // _In_
+                     const unsigned short w,                            // _In_
+                     const unsigned short h,                            // _In_
+                     HANDLE* const hInput,                              // _Out_
+                     HANDLE* const hOutput,                             // _Out_
+                     HANDLE* const hSignal,                             // _Out_
+                     PROCESS_INFORMATION* const piPty,                  // _Out_
+                     const EnvironmentVariableMapW& extraEnvVars = {}); // _In_
 
 bool SignalResizeWindow(const HANDLE hSignal,
                         const unsigned short w,
                         const unsigned short h);
 
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map);
+
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars);
+
+// Function Description:
+// - Updates an EnvironmentVariableMapW with the current process's unicode
+//   environment variables ignoring ones already set in the provided map.
+// Arguments:
+// - map: The map to populate with the current processes's environment variables.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map)
+{
+    LPWCH currentEnvVars{};
+    auto freeCurrentEnv = wil::scope_exit([&] {
+        if (currentEnvVars)
+        {
+            (void)FreeEnvironmentStringsW(currentEnvVars);
+            currentEnvVars = nullptr;
+        }
+    });
+
+    RETURN_IF_NULL_ALLOC(currentEnvVars = ::GetEnvironmentStringsW());
+
+    // Block is guaranteed to be double-NULL terminated at a minimum.
+    for (wchar_t const* lastCh{ currentEnvVars }; *lastCh != '\0';)

```

Comment:
I don't think I've ever seen a for loop initialized like this, I think
```suggestion
    for (wchar_t const* lastCh = currentEnvVars; *lastCh != '\0';)
```

would be prefered

Reply:
This is actually from the ISO C++ Guidelines and something I think you'll find more of then you might imagine in other Windows projects:

[ES.23: Prefer the {}-initializer syntax](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es23-prefer-the--initializer-syntax)

I tried to apply it to new functions I wrote for consistency.  If you'd still like me to change it, that's ok.

Reply:
fwiw I love the `{}`-initializer syntax.

Reply:
Alright sure, lets go for it. I don't really mind, and we should probably start conforming in places that we haven't been.